src/crimson/osd/shard_services.cc

```diff
@@ -97,10 +99,22 @@ seastar::future<> ShardServices::send_to_osd(
   } else {
     auto conn = cluster_msgr.connect(
         osdmap->get_cluster_addrs(peer).front(), CEPH_ENTITY_TYPE_OSD);
-    return conn->send(m);
+    return conn->send(std::move(m));
   }
 }
 
+seastar::future<> ShardServices::send_to_osd(
+  int peer, MessageRef m, epoch_t from_epoch) 
```

Comment:
Could you elaborate how much refactoring work is needed in OSD in order to remove this interface and `Connection::send(MessageRef)`?

Reply:
To drop `Connection::send(MessageRef)` we need to refactor the internals of Messenger (protocol for example). To drop `ShardServices::send_to_osd(MessageRef)` we need to refactor `PeeringState` which is currently shared between ceph-osd and crimson-osd because there is a method called `send_cluster_message()` which uses `send_to_osd()`.